### PR TITLE
[PT FE] Fix issue with aten::index with bool mask on axis

### PR DESCRIPTION
**Pull request by mvafin**
### Details:
 - *Fix issue with `aten::index` with bool mask on axis*

### Tickets:
 - *CVS-173495*

### CODE REVIEW

File: src/frontends/pytorch/src/utils.cpp [hunk 1]
```diff
@@ -797,8 +797,14 @@ bool index_tensor_on_list(ov::pass::NodeRegistry& rg,
         if (id_dtype == element::boolean || id_dtype == element::u8) {
             auto idx = rg.make<v0::Convert>(indices[i], element::u8);
             auto nonzero = rg.make<v3::NonZero>(idx);
+            Output<Node> masked_id;
+            if (indices.size() == 1) {
             auto input_order = rg.make<v0::Constant>(element::i32, Shape{2}, std::vector<int32_t>{1, 0});
```
Comment (rkazants) [inline src/frontends/pytorch/src/utils.cpp]:
```suggestion
                auto input_order = rg.make<v0::Constant>(element::i32, Shape{2}, std::vector<int32_t>{1, 0});
```

File: src/frontends/pytorch/src/utils.cpp [hunk 2]
```diff
@@ -797,8 +797,14 @@ bool index_tensor_on_list(ov::pass::NodeRegistry& rg,
...
-            auto masked_id = rg.make<v1::Transpose>(nonzero, input_order);
+            masked_id = rg.make<v1::Transpose>(nonzero, input_order);
```
Comment (rkazants) [inline src/frontends/pytorch/src/utils.cpp]:
```suggestion
                masked_id = rg.make<v1::Transpose>(nonzero, input_order);
```

File: tests/layer_tests/pytorch_tests/test_index.py [hunk 3]
```diff
@@ -74,6 +86,12 @@ def test_index_bool(self, input_shape, idx, case, ie_device, precision, ir_versi
         self._test(*self.create_model(case), ie_device, precision, ir_version,
                    kwargs_to_prepare_input={"input_shape": input_shape, "idx": idx})
 
+    @pytest.mark.nightly
+    @pytest.mark.precommit
+    def test_index_bool_with_axis(self, ie_device, precision, ir_version):
+        self._test(*self.create_model("bool_with_axis"), ie_device, precision, ir_version,
+                   kwargs_to_prepare_input={"input_shape": (2,2,5)}, trace_model=True)
```
Comment (mvafin) [inline tests/layer_tests/pytorch_tests/test_index.py]:
```suggestion
                   kwargs_to_prepare_input={"input_shape": (2, 2, 5)}, trace_model=True)
```

File: src/frontends/pytorch/src/utils.cpp [hunk 4]
```diff
@@ -797,8 +797,14 @@ bool index_tensor_on_list(ov::pass::NodeRegistry& rg,
...
-            auto input_order = rg.make<v0::Constant>(element::i32, Shape{2}, std::vector<int32_t>{1, 0});
-            auto masked_id = rg.make<v1::Transpose>(nonzero, input_order);
+            Output<Node> masked_id;
+            if (indices.size() == 1) {
+                auto input_order = rg.make<v0::Constant>(element::i32, Shape{2}, std::vector<int32_t>{1, 0});
+                masked_id = rg.make<v1::Transpose>(nonzero, input_order);
+            } else {
+                auto zero_const = rg.make<v0::Constant>(element::i32, Shape{1}, 0);
+                masked_id = rg.make<v0::Squeeze>(nonzero, zero_const);
+            }
```
Comment (mvafin) [inline src/frontends/pytorch/src/utils.cpp]:
```suggestion
            }
```

File: src/frontends/pytorch/src/utils.cpp [hunk 5]
```diff
@@ -797,8 +797,14 @@ bool index_tensor_on_list(ov::pass::NodeRegistry& rg,
...
             masked_indicies.push_back(masked_id);
```
Comment (bumbosiepsak) [inline src/frontends/pytorch/src/utils.cpp:808]: Nitpick: how about:
```
masked_indicies.push_back(std::move(masked_id));
```
?

File: src/frontends/pytorch/src/utils.cpp [hunk 6]
```diff
@@ -815,7 +821,7 @@ bool index_tensor_on_list(ov::pass::NodeRegistry& rg,
         return true;
     }
     // perform gather for single element case
-    if (advanced_ids.size() == 1) {
+    if (advanced_ids.size() == 1 && advanced_ids[0] == 0) {
```
Comment (bumbosiepsak) [inline src/frontends/pytorch/src/utils.cpp:824]: Nitpick: how about:
```
if (advanced_ids.size() == 1 && 0 == advanced_ids[0]) {
```
?

Rationale: protects against accidental typos like:
```
if (advanced_ids.size() == 1 && advanced_ids[0] = 0) {
```

File: tests/layer_tests/pytorch_tests/test_index.py [hunk 7]
```diff
@@ -9,9 +9,11 @@
 
 
 class TestIndex(PytorchLayerTest):
-    def _prepare_input(self, input_shape, idx):
+    def _prepare_input(self, input_shape, idx=None):
         import numpy as np
-        return (np.random.randn(*input_shape).astype(np.float32), idx)
+        if idx is not None:
```
Comment (bumbosiepsak) [inline tests/layer_tests/pytorch_tests/test_index.py]: Nitpick: how about:
```
input = np.random.randn(*input_shape).astype(np.float32)
return (input,) if idx is None else (input, idx)
```
?

Rationale:
- slightly less syntactic noise
- positive logic (`is None` vs `is not None`)

File: tests/layer_tests/pytorch_tests/test_index.py [hunk 8]
```diff
@@ -9,9 +9,11 @@
...
+            return (np.random.randn(*input_shape).astype(np.float32), idx)
```
Comment (bumbosiepsak) [inline tests/layer_tests/pytorch_tests/test_index.py]: Nitpick: stability/reproducibility of the test would increase if the testcase was fed with constant rather than random data.

File: tests/layer_tests/pytorch_tests/test_index.py [hunk 9]
```diff
@@ -35,11 +37,21 @@ class aten_index_getitem_bool(torch.nn.Module):
 
             def forward(self, x, idx):
                 return x.__getitem__(idx.to(torch.bool))
+
+        class aten_index_bool_with_axis(torch.nn.Module):
```
Comment (bumbosiepsak) [inline tests/layer_tests/pytorch_tests/test_index.py:34]: Question: is it needed/beneficial/performant to store the index? Or would the simpler version suffice:
```
class aten_index_bool_with_axis(torch.nn.Module):
    def forward(self, x):
        idx = torch.tensor([1, 0, 1, 0, 1], dtype=torch.bool)
        return x[:, :, idx]
```
?

File: tests/layer_tests/pytorch_tests/test_index.py [hunk 10]
```diff
@@ -35,11 +37,21 @@ class aten_index_getitem_bool(torch.nn.Module):
...
+            def __init__(self):
+                super().__init__()
+                self.idx = torch.tensor([1, 0, 1, 0, 1], dtype=torch.bool)
+
+            def forward(self, x):
+                return x[:,:,self.idx]
+
         cases = {
             "list": aten_index_list,
             "getitem": aten_index_getitem,
```
Comment (bumbosiepsak) [inline tests/layer_tests/pytorch_tests/test_index.py:44]: Nitpick: how about having regular/derived naming:
```
cases = {
    "list": aten_index_list,
    "list_with_bool": aten_index_list_with_bool,
    "getitem_with_bool": aten_index_getitem_with_bool,
    "bool_with_axis": aten_index_bool_with_axis,
}
```
?

Rationale:
- easier searching/grepping/refactoring
- fewer mappings in the head of someone reading this

File: tests/layer_tests/pytorch_tests/test_index.py [hunk 11]
```diff
@@ -35,11 +37,21 @@ class aten_index_getitem_bool(torch.nn.Module):
...
             "list_with_bool": aten_index_list_bool,
-            "getitem_with_bool": aten_index_getitem_bool
+            "getitem_with_bool": aten_index_getitem_bool,
+            "bool_with_axis": aten_index_bool_with_axis
```
Comment (bumbosiepsak) [inline tests/layer_tests/pytorch_tests/test_index.py]: Nitpick: add a coma after the last item:
```python
cases = {
    "list": aten_index_list,
    "getitem": aten_index_getitem,
    "list_with_bool": aten_index_list_bool,
    "getitem_with_bool": aten_index_getitem_bool,
    "bool_with_axis": aten_index_bool_with_axis, # Harmless coma here
}
```

for the sake of diffing/merging friendliness (when the testcases list keeps on growing):
```python
cases = {
    "list": aten_index_list,
    "getitem": aten_index_getitem,
    "list_with_bool": aten_index_list_bool,
    "getitem_with_bool": aten_index_getitem_bool,
    "bool_with_axis": aten_index_bool_with_axis, # Now this line doesn't generate a diff
    "fancy_future": aten_index_fancy_future
}
```

File: tests/layer_tests/pytorch_tests/test_index.py (hunk 9 above)
Comment (mvafin) [reply tests/layer_tests/pytorch_tests/test_index.py:34]: That may be sufficient, but lets try to make dummy models as close to real models as possible.

File: tests/layer_tests/pytorch_tests/test_index.py (hunk 11 above)
Comment (mvafin) [reply tests/layer_tests/pytorch_tests/test_index.py]:
```suggestion
            "bool_with_axis": aten_index_bool_with_axis,
```

File: tests/layer_tests/pytorch_tests/test_index.py [hunk 12]
```diff
@@ -9,9 +9,11 @@
 
 
 class TestIndex(PytorchLayerTest):
-    def _prepare_input(self, input_shape, idx):
+    def _prepare_input(self, input_shape, idx=None):
         import numpy as np
-        return (np.random.randn(*input_shape).astype(np.float32), idx)
+        if idx is not None:
+            return (np.random.randn(*input_shape).astype(np.float32), idx)
+        return (np.random.randn(*input_shape).astype(np.float32),)
```
Comment (mvafin) [inline tests/layer_tests/pytorch_tests/test_index.py]:
```suggestion
        x = np.random.randn(*input_shape).astype(np.float32)
        return (x,) if idx is None else (x, idx)
```

File: src/frontends/pytorch/src/utils.cpp (hunk 5 above)
Comment (mvafin) [reply src/frontends/pytorch/src/utils.cpp:808]: Maybe a good idea, but I the difference is low, so I would keep it as is for now

File: src/frontends/pytorch/src/utils.cpp (hunk 6 above)
Comment (mvafin) [reply src/frontends/pytorch/src/utils.cpp:824]: This reduce readability I think, and I don't think that will compile tbh

File: tests/layer_tests/pytorch_tests/test_index.py (hunk 10 above)
Comment (mvafin) [reply tests/layer_tests/pytorch_tests/test_index.py:44]: Don't think this will make any difference
